### PR TITLE
Ref #4426: Support CSimple expressions with all DSLs

### LINKED ISSUE #4426: Support CSimple expressions with all DSLs
So far we can only use CSimple expressions with the Java DSL and the XML DSL, the idea of this task is to propose a more generic approach in order to support all the existing DSL.

### PR DESCRIPTION
**Pull request by essobedo**
fixes #4426 

## Motivation

So far we can only use CSimple expressions with the Java DSL and the XML DSL, the idea of this task is to propose a more generic approach in order to support all the existing DSL.

## Modifications:

* Add the new extension `support-language` that does a dry run to collect all the expressions and produces `ExpressionBuildItem`
* Rewrite the way the CSimple expressions are extracted to leverage the result of the expression extractor
* Deprecate `CSimpleConfig` in favor of `ExpressionConfig`
* Add a Yaml file to the CSimple integration test to show that it also works with the Yaml DSL 
* Update the documentation about the limitations
* Add the ability to disable the expression extraction process
* Add a note and a warning indicating that `camel-csimple-joor` needs to be added to the classpath in case the expression extraction process failed or was disabled

## Result

The CSimple expressions can now be used with all DSLs, the only drawback with this new approach is the fact that it adds some log entries of type:
```
2023-01-19 16:36:31,779 INFO  [org.apa.cam.mai.MainSupport] (build-10) Expression Extractor 3.20.1 is starting
2023-01-19 16:36:32,214 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (build-10) Apache Camel 3.20.1 (Expression Extractor) is starting
2023-01-19 16:36:32,216 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (build-10) Routes startup (started:0)
2023-01-19 16:36:32,217 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (build-10) Apache Camel 3.20.1 (Expression Extractor) started in 128ms (build:22ms init:104ms start:2ms)
2023-01-19 16:36:32,218 INFO  [org.apa.cam.mai.MainSupport] (build-10) Waiting until complete: Duration max 1 seconds
2023-01-19 16:36:33,223 INFO  [org.apa.cam.mai.MainSupport] (build-10) Duration max seconds triggering shutdown of the JVM
2023-01-19 16:36:33,223 INFO  [org.apa.cam.mai.MainSupport] (build-10) Expression Extractor 3.20.1 shutdown
```

### CODE REVIEW

**Review by davsclaus — APPROVED** [review]

**Comment by davsclaus** [comment]
Nice to see the dry run as this can be used for other kind of optimizations

File: extensions-support/language/deployment/pom.xml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.camel.quarkus</groupId>
+        <artifactId>camel-quarkus-support-language-parent</artifactId>
+        <version>2.16.0-SNAPSHOT</version>
```
Comment (zhfeng) [inline extensions-support/language/deployment/pom.xml:25]: It should be `2.17-SNAPSHOT` now

File: extensions-support/language/pom.xml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.camel.quarkus</groupId>
+        <artifactId>camel-quarkus-extensions-support</artifactId>
+        <version>2.16.0-SNAPSHOT</version>
```
Comment (zhfeng) [inline extensions-support/language/pom.xml:25]: same as above

File: extensions-support/language/runtime/pom.xml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.camel.quarkus</groupId>
+        <artifactId>camel-quarkus-support-language-parent</artifactId>
+        <version>2.16.0-SNAPSHOT</version>
```
Comment (zhfeng) [inline extensions-support/language/runtime/pom.xml:25]: same as above

File: extensions-support/language/runtime/pom.xml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>camel-quarkus-support-language</artifactId>
+    <name>Camel Quarkus :: Support :: Language :: Runtime</name>
+
+    <properties>
+        <camel.quarkus.jvmSince>2.16.0</camel.quarkus.jvmSince>
+        <camel.quarkus.nativeSince>2.16.0</camel.quarkus.nativeSince>
```
Comment (zhfeng) [inline extensions-support/language/runtime/pom.xml:34]: 2.17.0

File: extensions-support/language/deployment/src/main/java/org/apache/camel/quarkus/support/language/deployment/LanguageSupportProcessor.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        }
+        final List<Class<?>> routeBuilderClasses = new ArrayList<>(routesBuilderClasses.size());
+        for (CamelRoutesBuilderClassBuildItem routesBuilderClass : routesBuilderClasses) {
+            final String className = routesBuilderClass.getDotName().toString();
+            final Class<?> cl = loader.loadClass(className);
+
+            if (RouteBuilder.class.isAssignableFrom(cl)) {
+                routeBuilderClasses.add(cl);
+            } else {
+                LOG.warnf("CSimple language expressions occurring in %s won't be compiled at build time", cl);
```
Comment (jamesnetherton) [inline extensions-support/language/deployment/src/main/java/org/apache/camel/quarkus/support/language/deployment/LanguageSupportProcessor.java:52]: Nitpick - Should we remove `CSimple` from the log message?

**Comment by jamesnetherton** [comment]
Something I just remembered about doing 'dry run' stuff at build time, is that you potentially run into issues once you start moving away from the 'happy path'. E.g if a `RouteBuilder` does some CDI injection, I think the build time Expression Extractor phase will fail.

Maybe we should add a note about that to the docs.

**Comment by essobedo** [comment]
> Something I just remembered about doing 'dry run' stuff at build time, is that you potentially run into issues once you start moving away from the 'happy path'. E.g if a `RouteBuilder` does some CDI injection, I think the build time Expression Extractor phase will fail.
> 
> Maybe we should add a note about that to the docs.

I see, I'm adding a fallback solution and a note about this limitation

**Comment by essobedo** [comment]
Note added and fallback solution proposed

**Review by zhfeng — APPROVED** [review]

**Review by jamesnetherton — APPROVED** [review]

**Comment by essobedo** [comment]
That's weird, the build complained about uncommitted files that are not even related to these changes (cxf and controlbus-statistics).

**Pull request merged**